test/benchmarks/external-setup.sh

```diff
@@ -0,0 +1,104 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# solidity is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with solidity.  If not, see <http://www.gnu.org/licenses/>
+#
+# (c) 2024 solidity contributors.
+#------------------------------------------------------------------------------
+
+set -euo pipefail
+
+repo_root=$(cd "$(dirname "$0")/../../" && pwd)
+BENCHMARK_DIR="${BENCHMARK_DIR:-${repo_root}/benchmarks}"
+
+function neutralize_version_pragmas {
+    find . -name '*.sol' -type f -print0 | xargs -0 \
+        sed -i -E -e 's/pragma solidity [^;]+;/pragma solidity *;/'
+}
+
+mkdir -p "$BENCHMARK_DIR"
+cd "$BENCHMARK_DIR"
+
+if [[ ! -e openzeppelin/ ]]; then
+    git clone --depth=1 https://github.com/OpenZeppelin/openzeppelin-contracts openzeppelin/ --branch v5.0.2
+    pushd openzeppelin/
+    forge install
+    popd
+else
+    echo "Skipped openzeppelin/. Already exists."
+fi
+
+if [[ ! -e uniswap-v4/ ]]; then
+    git clone --single-branch https://github.com/Uniswap/v4-core uniswap-v4/
+    pushd uniswap-v4/
+    git checkout ae86975b058d386c9be24e8994236f662affacdb # branch main as of 2024-06-06
+    forge install
+    popd
+else
+    echo "Skipped uniswap-v4/. Already exists."
+fi
+
+if [[ ! -e seaport/ ]]; then
+    git clone --single-branch https://github.com/ProjectOpenSea/seaport
+    pushd seaport/
+    # NOTE: Can't select the tag with `git clone` because a branch of the same name exists.
+    git checkout tags/1.6
+    forge install
+    neutralize_version_pragmas
+    popd
+else
+    echo "Skipped seaport/. Already exists."
+fi
+
+if [[ ! -e eigenlayer/ ]]; then
+    git clone --depth=1 https://github.com/Layr-Labs/eigenlayer-contracts eigenlayer/ --branch v0.3.0-holesky-rewards
+    pushd eigenlayer/
+    forge install
+    popd
+else
+    echo "Skipped eigenlayer/. Already exists."
+fi
+
+if [[ ! -e sablier-v2/ ]]; then
+    git clone --depth=1 https://github.com/sablier-labs/v2-core sablier-v2/ --branch v1.1.2
+    pushd sablier-v2/
+    # NOTE: To avoid hard-coding dependency versions here we'd have to install them from npm
+    forge install --no-commit \
+        foundry-rs/forge-std@v1.5.6 \
+        OpenZeppelin/openzeppelin-contracts@v4.9.2 \
+        PaulRBerg/prb-math@v4 \
+        PaulRBerg/prb-test@0.6.4 \
+        evmcheb/solarray@a547630 \
+        Vectorized/solady@v0.0.129
+   cat <<EOF > remappings.txt
+@openzeppelin/contracts/=lib/openzeppelin-contracts/contracts/
+forge-std/=lib/forge-std/
+@prb/math/=lib/prb-math/
+@prb/test/=lib/prb-test/
+solarray/=lib/solarray/
+solady/=lib/solady/
+EOF
```

Comment:
I wanted to avoid using npm here since `package.json` can have scripts that run automatically, including ones that compile the project and I don't want this to devolve into mountains of file patching like it did in external tests.

I'm on the fence whether to stick to Foundry-based projects or add support for Hardhat-based ones too though. I wanted to avoid it originally, because Foundry makes our use case much simpler (`forge install` + `forge build` with the right options compared to tons of search&replace in config files we need to configure a Hardhat project), but there are a few Hardhat-based ones that would be really good to have as benchmarks. And if we did add them, using npm here would not make things significantly worse than they would be already.

Reply:
I'm wondering if we shouldn't try to reuse the python test suite for our benchmarks. For example, we could use the `FoundryRunner` for each project we want to benchmark, and then define a `run_benchmark` function in `runners/base.py` or another appropriate location. This could also eliminate the duplication of `benchmark_contract` function, which is quite similar in both `local.sh` and `external.sh`. We could also move the `bytecode_size` function to a common location. The idea would be something like:

```python
from runners.base import run_benchmark
from runners.base import TestConfig
from runners.foundry import FoundryRunner, run_forge_command
from test_helpers import replace_version_pragmas, download_project

class OpenZeppelinRunner(FoundryRunner):
    def configure(self):
        # We could download the project here by calling `download_project`
        # like you are doing for each project, or let the `run_benchmark` function do it.
        # download_project(self.test_dir, self.config.repo_url, self.config.ref)

        # Override the configure method to skip `setup_presets_profiles` which is
        # called by default in the base class.
        run_forge_command("forge install", self.env)

        # We could also call `replace_version_pragmas` or other setup functions
        # replace_version_pragmas(self.test_dir)

test_config = TestConfig(
    name="OpenZeppelin",
    repo_url="https://github.com/OpenZeppelin/openzeppelin-contracts.git",
    ref="v5.0.2",
)

sys.exit(run_benchmark(OpenZeppelinRunner(argv=sys.argv[1:], config=test_config)))
```

Alternatively, we could reuse the setup free functions in `test_helpers.py`, as the runner might require extensive overrides.

Reply:
I mean, I'm fine with doing it in shellscript as well. I'm just throwing a possible idea here.
So maybe we could just merge this in the way it is and think about refactoring it to python later ;)

Reply:
You do have a point. The more I tweak it the more things I find that complicate it. The need to hard-code dependencies for Sablier is one thing. Another I just found is that I actually need to patch `foundry.toml` to remove any `via_ir = true` because then it does not build legacy by default. I was getting wrong legacy results for Uniswap because of that. If the script needs any more of such mechanisms we're probably better off redoing it on top of your Python scripts. That actually sounds like a much better way to support Hardhat based projects here than what I was thinking of initially.

I'd rather not do that right away though. I put this together quickly and I really didn't want to spend too much time on it, because it's a distraction from the main task. I based it on Nikola's script because it was self-contained, easy to adapt and had all the features I needed. I was also able to reuse bits from [my test setup in `parasolc`](https://github.com/cameel/parasolc/tree/master/test). The Python scripts are still missing some things I needed like timing or the ability to reuse a checkout so that route seemed like more work. I'm open to refining it later, but for now I'd rather just merge it in the Bash form and be done with it :)

> However, it seems that something has broken the Python setup on the Windows CI.

I don't think it's related to the script. It looked like a failure to download or install the python package. I suspect it will go away when I rerun.

Reply:
Fair enough. So my only remaining question is if we should move `bytecode_size` function to common scripts, since it is duplicated there. Other than that, I'm fine with the PR ;)